### ISSUE TITLE
Implement InfoNCE loss from scratch with learnable temperature

### ISSUE THREAD

**Issue by nlklfor** [nlklfor/dl-multimodal-semantic-search#5]
## Task
Implement symmetric InfoNCE in `src/loss/infonce.py`. Temperature must be a learnable `nn.Parameter` initialized at `log(1/0.07)`.

## Acceptance criteria
- [x] Loss is symmetric (image→text AND text→image averaged)
- [x] Temperature is learnable by default, fixed when float is passed
- [x] Perfect alignment (identical embeddings) → loss ≈ 0
- [x] Random embeddings → loss ≈ log(batch_size)
- [x] Gradients flow through both input tensors
- [x] All tests in `tests/test_loss.py` pass

## Key formula
```python
logits = (image_emb @ text_emb.T) / temperature   # (B, B)
labels = torch.arange(B)
loss = (cross_entropy(logits, labels) + cross_entropy(logits.T, labels)) / 2
```

**Comment by nlklfor** [comment]
✅ Implemented by @muhammadov-q as part of the encoder work. Code reviewed and tested — all 11 unit tests passing. Thanks!
